batch-processing/helpers.go

```diff
@@ -0,0 +1,185 @@
+package main
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/csv"
+	"encoding/hex"
+	"fmt"
+	"image"
+	_ "image/gif"
+	_ "image/jpeg"
+	_ "image/png"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"errors"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/cenkalti/backoff/v4"
+)
+
+const (
+	InvalidCSVFormat   = "csv file must have a header row with 'url' as the first column"
+	CouldNotFetchImage = "received status %d when trying to download image"
+	EmptyCSV           = "provided CSV file appears to be empty"
+)
+
+/**
+* Download the image from the URL
+* @param: {string} URL - the URL of the image to download
+* @return: {io.Reader} body - the body of the image
+* @return: {string} ext - the file extension (format) of the image
+* @return: {error} err - any error that occurred
+ */
+func DownloadFileFromUrl(URL string) (io.Reader, string, [16]byte, error) {
+	response, err := http.Get(URL)
+	hash := [16]byte{}
+	if err != nil {
+		return nil, "", hash, err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
+		return nil, "", hash, fmt.Errorf(CouldNotFetchImage, response.StatusCode)
+	}
+
+	var buf bytes.Buffer
+
+	tee := io.TeeReader(response.Body, &buf)
+
+	// this reads only part of the file for format etc
+	_, format, err := image.DecodeConfig(tee)
+
+	if err != nil {
+		return nil, "", hash, err
+	}
+	// copy rest of the file
+	io.Copy(&buf, tee)
```

Comment:
This buffers the whole file in memory - can you think of a way to avoid doing so?

Reply:
I have fixed it now (I think), but it's not as elegant as I'm having to rename the file after I know the checksum. 

Is this ok, or am I missing a better way?

Reply:
This is pretty cute!

Another way, which could avoid the rename, would be to use a [`MultiReader`](https://pkg.go.dev/io#MultiReader), where you could make a `TeeReader` as you were before to output to the image decoder and keep the rest as a buffer, then set up a `MultiReader` with (what you already read, what you've not yet read).